File: CellUniverse/CellUniverse.Models/TheGameOfLife/Universe.cs
using System;
using System.Collections.Generic;

namespace CellUniverse.Models.TheGameOfLife {

    using Infrastructure.Interfaces;

    internal sealed class Universe : IUniverseModel {

        private int virtualWidth, virtualHeight;
        private Cell[] generationPlacement;

        private ComputeScheduler scheduler;

        private Random random;

        IEnumerable<Tuple<int, int, bool>> IUniverseModel.NextGeneration {
            get { return scheduler.NextGeneration; }
        }

        public Universe(int width, int height) {
            Initialize(width, height);
            FillRandom();
            BindAllToContainer(generationPlacement);
            scheduler.Start();
        }

        public Universe(bool[,] generation) {
            Initialize(generation.GetLength(0), generation.GetLength(1));
            ArrayCopy(generation, generationPlacement);
            BindAllToContainer(generationPlacement);
            scheduler.Start();
        }

        private void Initialize(int width, int height) {

            virtualWidth = width;
            virtualHeight = height;

            generationPlacement = GetNewEmptyUniverse(virtualWidth * virtualHeight);
            scheduler = new ComputeScheduler(generationPlacement, 1);

            random = new Random();
        }

        private void FillRandom() {
            for (int i = 0; i < generationPlacement.Length; i++) {
                generationPlacement[i] = random.Next(2) == 0;
            }
        }

        private void ArrayCopy(bool[,] source, Cell[] destinalion) {

            for (int x = 0; x < virtualWidth; x++) {
                for (int y = 0; y < virtualHeight; y++) {
                    destinalion[x + y * virtualWidth] = source[x, y];
                }
            }
        }

        private void BindAllToContainer(Cell[] container) {
            for (int x = 0; x < virtualWidth; x++) {
                for (int y = 0; y < virtualHeight; y++) {
                    container[x + y * virtualWidth]
                        .ConnectToNeighbours(x, y, container, virtualWidth, virtualHeight);
                }
            }
        }

        private Cell[] GetNewEmptyUniverse(int length) {
            Cell[] resultUniverse = new Cell[length];
            for (int i = 0; i < length; i++) {
                resultUniverse[i] = false;
            }
            return resultUniverse;
        }
    }
}

File: CellUniverse/CellUniverse.Models.Native/Universe.cpp
#include <random>
#include "Universe.h"

namespace CellUniverse {
	namespace Models {
		namespace Native {
			namespace CPP {

				int CUniverse::countUniverses = 0;

				CUniverse::CUniverse(int width, int height) {
					this->virtualWidth = width;
					this->virtualHeight = height;
					universePlacement = new bool[width * height];
					FillRandom();
					scheduler = new CUDA::CComputeScheduler(universePlacement, virtualWidth, virtualHeight, 8);
					scheduler->Start();
					++countUniverses;
				}

				void CUniverse::FillRandom() {
					std::srand(countUniverses);
					int index = virtualWidth * virtualHeight;
					while (index--) {
						universePlacement[index] = std::rand() % 2 != 0;
					}
				}

				bool* CUniverse::GetNext() {
					scheduler->GetNextGeneration(universePlacement);
					return universePlacement;
				}

				CUniverse::~CUniverse() {
					delete scheduler;
					delete universePlacement;
					universePlacement = nullptr;
					virtualWidth = virtualHeight = 0;
				}
			}
		}
	}
}